src/compiler/checker.ts

```diff
@@ -4845,6 +4861,92 @@ namespace ts {
+         */
+        function getDefaultFromTypeParameter(typeParameter: TypeParameter): Type | undefined {
+            if (!typeParameter.default) {
+                if (typeParameter.target) {
+                    const targetDefault = getDefaultFromTypeParameter(typeParameter.target);
+                    typeParameter.default = targetDefault ? instantiateType(targetDefault, typeParameter.mapper) : noConstraintOrDefaultType;
+                }
+                else {
+                    const defaultDeclaration = typeParameter.symbol && forEach(typeParameter.symbol.declarations, decl => isTypeParameter(decl) && decl.default);
+                    typeParameter.default = defaultDeclaration ? getTypeFromTypeNode(defaultDeclaration) : noConstraintOrDefaultType;
+                }
+            }
+            return typeParameter.default === noConstraintOrDefaultType ? undefined : typeParameter.default;
+        }
+
+        /**
+         * Gets the default type for a type parameter.
+         *
+         * If the type parameter is the result of an instantiation, this gets the instantiated
+         * default type of its target. If the type parameter has no default type, or if the default
+         * type circularly references the type parameter, `undefined` is returned.
+         *
+         * This function *does* perform a circularity check.
+         */
+        function getDefaultOfTypeParameter(typeParameter: TypeParameter): Type | undefined {
+            return hasNonCircularDefault(typeParameter) ? getDefaultFromTypeParameter(typeParameter) : undefined;
+        }
+
+        /**
+         * Determines whether a type parameter has a non-circular default type.
+         *
+         * Note that this function also returns `true` if a type parameter *does not* have a
+         * default type.
+         */
+        function hasNonCircularDefault(typeParameter: TypeParameter): boolean {
+            return getResolvedDefault(typeParameter) !== circularConstraintOrDefaultType;
+        }
+
+        /**
+         * Resolves the default type of a type parameter.
+         *
+         * If the type parameter has no default, the `noConstraintOrDefaultType` singleton is
+         * returned. If the type parameter has a circular default, the
+         * `circularConstraintOrDefaultType` singleton is returned.
+         */
+        function getResolvedDefault(typeParameter: TypeParameter): Type {
+            if (!typeParameter.resolvedDefault) {
+                if (!pushTypeResolution(typeParameter, TypeSystemPropertyName.ResolvedDefault)) {
+                    return circularConstraintOrDefaultType;
+                }
+                const defaultType = getDefaultFromTypeParameter(typeParameter);
+                const type = defaultType && getResolvedDefaultWorker(defaultType);
+                if (!popTypeResolution()) {
+                    return typeParameter.resolvedDefault = circularConstraintOrDefaultType;

```

Comment:
I removed this following our discussion regarding circular references here. Instead, now when a type parameter default points to an earlier type parameter, it gets the type argument supplied for that parameter or the default selected for that type parameter. If the type parameter default points to a later type parameter it becomes `emptyObjectType`